drift/core/psmc.py

```diff
@@ -201,6 +206,356 @@ def _work_fisher_bias_m(self, mi):
+            print("Fisher matrix file: %s exists. Skipping..." % ffile)
+            return
+
+        mpiutil.barrier()
+
+        # Pre-compute all the angular power spectra for the bands.
+        self.genbands()
+
+        comm = MPI.COMM_WORLD
+        size = comm.Get_size()
+        rank = comm.Get_rank()
+
+        # Split up all m-modes such that each MPI process receives one m-mode
+        m_chunks, low_bound, upp_bound = self.split_single(
+            (self.telescope.mmax + 1), size
+        )
+        num_chunks = m_chunks.shape[0]
+
+        n = self.nsamples
+
+        # Option to add another dimension for a noise term
+        nbands = self.nbands + 1 if self.noise else self.nbands
+
+        # Create an MPI array for the q-estimator distributed over bands
+        self.qa = mpiarray.MPIArray(
+            (self.telescope.mmax + 1, nbands, n),
+            axis=1,
+            comm=MPI.COMM_WORLD,
+            dtype=np.float64,
+        )
+
+        self.qa[:] = 0.0
+
+        # This loop is designed such that the total number of m's is split up in chunks.
+        # Each chunk contains as many m's as there are ranks such that each rank gets exactly one m-mode.
+        for ci, num_m in enumerate(m_chunks):
+            if mpiutil.rank0:
+                print("Starting chunk %i of %i" % (ci, num_chunks))
+
+            loc_num, loc_start, loc_end = mpiutil.split_local(num_m)
+            mi = np.arange(low_bound[ci], upp_bound[ci])[loc_start:loc_end]
+
+            # At last iteration of m's, it is necessary to assign m's to remaining ranks otherwise the MPI communication crashes.
+            if len(mi) == 0:
+                mi = np.array([low_bound[ci] + rank])
+
+            if len(mi) != 0:
+                print("Processing m-mode %i on rank %i" % (mi, rank))
+
+            nfreq = self.telescope.nfreq
+            lmax = self.telescope.lmax
+
+            if loc_num > 0:
+                if self.num_evals(mi) > 0:
+                    # Generate random KL data
+                    x = self.gen_sample(mi, n)
+                    vec1 = self.project_vector_kl_to_sky(mi, x)
+                    # Make array contiguous
+                    vec1 = np.ascontiguousarray(vec1.reshape(nfreq, lmax + 1, n))
+
+                # If I don't have evals - return zero vector. Each MPI process must have some data to pass around. Otherwise send and receive won't work.
+                else:
+                    vec1 = np.zeros((nfreq, lmax + 1, n), dtype=np.complex128)
+
+            else:
+                vec1 = np.zeros((nfreq, lmax + 1, n), dtype=np.complex128)
+
+            st_q = time.time()
+
+            # Loop over total number of ranks given by `size`.
+            for ir in range(size):
+                # Only fill qa if we haven't reached mmax
+                if mi < (self.telescope.mmax + 1):
+                    self.q_estimator(mi, vec1)
+
+                    # TO DO: Calculate q_a for noise power (x0^H N x0 = |x0|^2)
```

Comment:
I added her a TODO item - at the moment we have no capability of simulating a noise bias. 
On a note: 
- the original psestimation code doesn't estimate a noise bias for a auto power spectrum (why?)
-  Is x0^2 (=normalized KL-modes) supposed to be the noise bias? According to the m-mode paper see equation (91) -  I thought you need to sandwich the estimator between realizations of white noise to get b_a.

Reply:
oh yeah, and the only that noise is set to True is in the Crosspower class and there it should be zero, so I don't see the point why it's even calculated?!

Reply:
Comment to self and anyone else reading: We won't save noise in KL-basis to q-estimator. IMO it really doesn't need to be there.

Reply:
I'm confused by these comments. Isn't the noise bias estimated by the combination of `fisher_bias` and `gen_sample`?

I guess the problem is that the `noiseonly` argument isn't set to `True` when `zero_mean == False` which it should be?